src/nnet3/nnet-chain-training.cc

```diff
@@ -68,23 +68,49 @@ void NnetChainTrainer::Train(const NnetChainExample &chain_eg) {
                              &request);
   const NnetComputation *computation = compiler_.Compile(request);
 
+  // no adversarial training on the first minibatch to avoid some stats to be
```

Comment:
I really want this code to be clear, and for the "simple path" (no adversarial training) not to be cluttered by the adversarial-training code.  I think it would be clearer if you were to create a function like this [and similar in the non-chain code]:
```
void NnetChainTrainer::TrainInternal(const NnetChainExample &eg,
                                                            const NnetComputation &computation,
                                                            bool is_adversarial_step);
```
You could call FreezeNaturalGradient() outside of that function, so that it would
have the same code in both cases (it would just have to pass 'is_adversarial_step' in).

BTW, regarding the asserts that failed before you disabled the 1st-minibatch adversarial step: we should probably remove those asserts and make sure the code does something reasonable when the stats are 
negative.  Because even disabling the 1st-minibatch adversarial step they could still get negative, due to being scaled by the max-change code.  The self-repair code could just return if the counts are negative.  Before calling the fast-l LSTM kernels you just need to make sure the counts are nonzero to prevent division by zero; the actual value hardly matters at all as this condition is so rare and the self-repair is so gentle.